### PR TITLE
👷 K6

### PR DESCRIPTION
**Pull request by wdbasson**
### **K6 Load Testing Scripts**
⚠️ **This is still a WIP. Refactoring and standardisation pending....**

Initial load testing scripts and GH Actions pipeline config to run K6 in parallel with K8s Pytests. Making it part of the pipeline to continuously validate the code and avoid drift.

- Split pipeline to differentiate between K8s deployment and test jobs
- Initial K6 flows:
  - setup function to create schema, issuer and cred def
  - create holders
  - establish connections
  - issue credentials
  - proof requests

### CODE REVIEW

File: .github/workflows/continuous-deploy.yml [hunk 1]
```diff
@@ -839,3 +924,82 @@ jobs:
...
+          -e GOVERNANCE_CLIENT_ID=${{ secrets.GOVERNANCE_CLIENT_ID }} \
+          -e CLIENT_SECRET=${{ secrets.CLIENT_SECRET }} \
+          -e GOVERNANCE_CLIENT_SECRET=${{ secrets.GOVERNANCE_CLIENT_SECRET }} \
+          -e CLOUDAPI_URL=${{ secrets.CLOUDAPI_URL }} \
+          -e OAUTH_ENDPOINT=${{ secrets.OAUTH_ENDPOINT }} \
+          -e GOVERNANCE_OAUTH_ENDPOINT=${{ secrets.GOVERNANCE_OAUTH_ENDPOINT }} \
+          --workdir /scripts \
+          --entrypoint /bin/sh \
+          ghcr.io/${{ github.repository_owner }}/xk6:${{ steps.meta.outputs.version }} \
+          /scripts/run_tests.sh
```
Comment (rblaine95) [inline .github/workflows/continuous-deploy.yml:1004]:
```suggestion
            -v ${{ github.workspace }}/scripts/k6:/scripts \
            -e CLIENT_ID=${{ secrets.CLIENT_ID }} \
            -e GOVERNANCE_CLIENT_ID=${{ secrets.GOVERNANCE_CLIENT_ID }} \
            -e CLIENT_SECRET=${{ secrets.CLIENT_SECRET }} \
            -e GOVERNANCE_CLIENT_SECRET=${{ secrets.GOVERNANCE_CLIENT_SECRET }} \
            -e CLOUDAPI_URL=${{ secrets.CLOUDAPI_URL }} \
            -e OAUTH_ENDPOINT=${{ secrets.OAUTH_ENDPOINT }} \
            -e GOVERNANCE_OAUTH_ENDPOINT=${{ secrets.GOVERNANCE_OAUTH_ENDPOINT }} \
            --workdir /scripts \
            --entrypoint /bin/sh \
            ghcr.io/${{ github.repository_owner }}/xk6:${{ steps.meta.outputs.version }} \
            /scripts/run_tests.sh
```

File: .github/workflows/continuous-deploy.yml [hunk 2]
```diff
@@ -538,6 +538,91 @@ jobs:
...
+      id-token: write # Required to authenticate with AWS
+      contents: read # Required to clone this repository
+      checks: write # Required for action-junit-report
+      pull-requests: write # Required to comment on PRs for Pytest coverage comment
+    runs-on: ubuntu-latest
+
+    timeout-minutes: 35
+
+    env:
+      TAILSCALE_VERSION: 1.68.1
```
Comment (rblaine95) [inline .github/workflows/continuous-deploy.yml:558]:
```suggestion
      TAILSCALE_VERSION: 1.70.0
```

File: .github/workflows/continuous-deploy.yml [hunk 3]
```diff
@@ -538,6 +538,91 @@ jobs:
...
+      HELMFILE_VERSION: v0.166.0
+      HELM_VERSION: v3.15.2
```
Comment (rblaine95) [inline .github/workflows/continuous-deploy.yml:560]:
```suggestion
      HELM_VERSION: v3.15.3
```

**Review by rblaine95 — CHANGES_REQUESTED** [review]

File: scripts/k6/README.md
Comment (ff137) [inline scripts/k6/README.md:1]: note: readme is an empty file

File: scripts/k6/tenant.js [hunk 4]
```diff
@@ -0,0 +1,809 @@
...
+  // if number equals 1, sleep for 1 second
+  // if (random === 1) {
+  //   console.log("Sleeping for 1 second");
+  //   eventReceived = false
+  // }
+  return eventReceived;
+}
+
+export function getCredentialDefinitionId(bearerToken, issuerAccessToken, credDefTag) {
+  const url = `${__ENV.CLOUDAPI_URL}/tenant/v1/definitions/credentials?schema_version=0.1.0`;
```
Comment (ff137) [inline scripts/k6/tenant.js:476]: minor note: this schema_version=0.1.0 seems it should use the SCHEMA_VERSION env var, which equals 0.1.0

File: .github/workflows/continuous-deploy.yml [hunk 5]
```diff
@@ -538,6 +538,91 @@ jobs:
...
+          ref: ${{ env.branch_name }}
+
+      - name: Sops Binary Installer
+        uses: mdgreenwald/mozilla-sops-action@v1.6.0
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v4
+        with:
+          aws-region: af-south-1
+          role-to-assume: arn:aws:iam::402177810328:role/cicd
```
Comment (ff137) [inline .github/workflows/continuous-deploy.yml:614]: I assume this role doesn't need to be a github secret?

**Review by ff137 — APPROVED** [review]
lots of js code, but reviewed the k6 logs and they look good 👍  nice work!

File: .github/workflows/continuous-deploy.yml (hunk 5 above)
Comment (rblaine95) [reply .github/workflows/continuous-deploy.yml:614]: AWS Account ID isn't considered sensitive

File: scripts/k6/README.md
Comment (wdbasson) [reply scripts/k6/README.md:1]: Pushed in latest commit. It is just a disclaimer that this is WIP.

File: scripts/k6/tenant.js (hunk 4 above)
Comment (wdbasson) [reply scripts/k6/tenant.js:476]: It should be passed to the function. Hardcoded the schema stuff before when it was still being created outside of k6. Will fix that as I add flows for schema creation.

File: scripts/k6/README.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
+### K6 Load Testing Scripts
+⚠️ Discalimer: This is a work in progress. Refactoring and standardisation pending... ⚠️
```
[the remark this answers is not in the document]
Comment (ff137) [reply scripts/k6/README.md:2]: @wdbasson PS, the vscode extension `markdownlint` can autofix some of these warnings. and `Code Spell Checker` is handy for catching typos, like Discalimer ;-) no biggie, but the extensions are worth sharing

File: .github/workflows/continuous-deploy.yml [hunk 7]
```diff
@@ -318,15 +318,12 @@ jobs:
       pull-requests: write # Required to comment on PRs for Pytest coverage comment
     runs-on: ubuntu-latest
 
-    timeout-minutes: 35
+    timeout-minutes: 10
 
     env:
       TAILSCALE_VERSION: 1.68.1
```
Comment (rblaine95) [inline .github/workflows/continuous-deploy.yml:324]:
```suggestion
      TAILSCALE_VERSION: 1.70.0
```

File: .github/workflows/continuous-deploy.yml [hunk 8]
```diff
@@ -318,15 +318,12 @@ jobs:
...
       HELMFILE_VERSION: v0.166.0
       HELM_VERSION: v3.15.2
```
Comment (rblaine95) [inline .github/workflows/continuous-deploy.yml:326]:
```suggestion
      HELM_VERSION: v3.15.3
```

**Review by rblaine95 — CHANGES_REQUESTED** [review]

File: .github/workflows/continuous-deploy.yml (hunk 7 above)
Comment (rblaine95) [reply .github/workflows/continuous-deploy.yml:324]: We should move these env vars to be pipeline wide

**Review by rblaine95 — APPROVED** [review]

**Pull request merged**